src/main/java/net/imagej/ops/OpListing.java

```diff
@@ -0,0 +1,348 @@
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import org.scijava.ItemIO;
+import org.scijava.module.DefaultMutableModuleItem;
+import org.scijava.module.ModuleInfo;
+import org.scijava.module.ModuleItem;
+import org.scijava.module.MutableModuleItem;
+import org.scijava.util.Types;
+
+/**
+ * A data structure designed to capture an abstract algorithm. As far as ImageJ
+ * Ops is concerned, Ops that have the same {@link OpListing} are <em>different
+ * implementations of the same algorithm</em>, much as method overloading in
+ * Java might be considered different implementations of the same functionality.
+ * <p>
+ * If you have <b>concrete Objects</b> instead of <b>types</b>, use
+ * {@link OpRef}.
+ *
+ * @author Gabriel Selzer
+ * @see OpRef
+ * @see OpInfo
+ */
+public class OpListing {
+
+	/** The name of the Op */
+	private final String name;
+	/** The fundamental type of the Op */
+	private final Class<?> functionalType;
+	/** The inputs expected by the Op */
+	private final List<String> inputNames;
+	private final List<Type> inputTypes;
+	/** The outputs given by the Op */
+	private final Optional<String> returnName;
+	private final Optional<Type> returnType;
+
+	/**
+	 * Convenience constructor, mainly used by {@link OpInfo}s.
+	 *
+	 * @param info the {@link OpInfo} to scrape for the listing data.
+	 */
+	public OpListing(final OpInfo info) {
+		this.name = info.getName();
+		this.functionalType = info.getType();
+		this.inputNames = new ArrayList<>();
+		this.inputTypes = new ArrayList<>();
+		for (final ModuleItem<?> input : info.inputs()) {
+			inputNames.add(input.getName());
+			inputTypes.add(input.getGenericType());
+		}
+		final Optional<ModuleItem<?>> first = info.outputs().stream().filter(
+			item -> !item.isInput()).findFirst();
+		this.returnName = first.map(ModuleItem::getName);
+		this.returnType = first.map(ModuleItem::getGenericType);
+	}
+
+	/**
+	 * Convenience {@link OpListing} constructor for Ops without returns.
+	 *
+	 * @param name the name of the Op.
+	 * @param functionalType the basic type of the Op (e.g.
+	 *          {@link net.imagej.ops.special.function.UnaryFunctionOp})
+	 * @param inputNames the input names of the Op
+	 * @param inputTypes the input types of the Op
+	 */
+	public OpListing( //
+		final String name, //
+		final Class<?> functionalType, //
+		final List<String> inputNames, //
+		final List<Type> inputTypes)
+	{
+		this(name, functionalType, inputNames, inputTypes, null, null);
+	}
+
+	/**
+	 * Standard {@link OpListing} constructor.
+	 *
+	 * @param name the name of the Op.
+	 * @param functionalType the basic type of the Op (e.g.
+	 *          {@link net.imagej.ops.special.function.UnaryFunctionOp})
+	 * @param inputNames the input names of the Op
+	 * @param inputTypes the input types of the Op
+	 * @param returnName the return name of the Op
+	 * @param returnType the return type of the Op
+	 */
+	public OpListing( //
```

Comment:
The only code calling this constructor is in this class, so maybe we shouldn't expose this publicly yet.

Reply:
Made private in 64aae77d1. I'd actually prefer that people use this constructor, but we can always make it public later.